tethys_apps/base/app_base.py

```diff
@@ -238,22 +238,22 @@ def register_url_maps(self):
 
     @property
     def registered_url_maps(self):
+        # TODO remove deprecation code
+        try:
+            self._registered_url_maps = self.url_maps()
+            from tethys_cli.cli_colors import write_warning
+            write_warning(
+                f'Deprecation Warning: The "{self.name}" app has the `url_maps` method defined. '
+                'This method is now deprecated. '
+                'Please use the new "controller", "consumer", and "handler" decorators to register URL maps '
+                '(see docs: http://docs.tethysplatform.org/en/stable/tethys_sdk/routing.html).'
+            )
+        except AttributeError:
+            pass
+
         if self._registered_url_maps is None:
             self._registered_url_maps = self.register_url_maps()
```

Comment:
Can we add an optional argument to the ``regiester_url_maps`` method that will allow passing a list of UrlMaps to initialize it with when overriding. Example usage:

```python
def register_url_maps(self):

    url_maps = [...]

    url_maps = super().register_url_maps(url_maps)

    url_maps.append(UrlMap(...))
    return url_maps
```

Reply:
@swainn What are you expecting the `register_url_maps` to do with the list of `UrlMap`s that you pass to it? In you're example you pass some in to the `super` call, and then also append some afterward. So I'm assuming there is something different about the once you pass vs the ones you append after.

Reply:
In AGWA we ran into an edge case. The app defines no URL endpoints itself with the controller decorator but instead uses ATCore to generate its UrlMaps. The endpoint used as the home view is included in these generated URLs. The problem is that the `register_url_maps` method looks for a controller matching the `index` property and fails if it is not found.

We were able to work around this issue by not calling the super method in our override. However, if AGWA does end up implementing its own controllers in the future with the controller decorator, we'd have no way to load them, since if we call the super method, it errors out.

My suggestion is to allow seeding the `register_url_maps` method with manually created UrlMaps. Another alternative I just thought of is that we have an optional parameter for specifying only the home UrlMap for cases when it is manually defined and not decorated.

Reply:
I think that you mentioned that your home controller is actually provided by an extension, so I'm guessing that's the reason for wanting to pass a list of maps? And adding extra controller modules via the app class `controller_modules` doesn't work?

Here is the actual implementation of the `register_url_maps`:

```python
        from .controller import register_controllers

        controller_modules = [f'{self.package_namespace}.{self.package}.{module_name}' for module_name in
                              set(DEFAULT_CONTROLLER_MODULES + self.controller_modules)]
        return register_controllers(
            root_url=self.root_url,
            modules=controller_modules,
            index=self.index,
            catch_all=self.catch_all,
        )
``` 

The only way for there to be any value in passing anything to this method would be if we also pass it through to the `register_controllers` method. The `register_cpontrollers` method actually creates 'UrlMaps` so it may be best to just pass in the `UrlMap` kwargs.

Implementation would looks something like this:


App Class Override:

```python
def register_url_maps(self):

    controller_kwargs = [
        dict(
            name='url_name',
            url='final_url',
            controller='controller',
            protocol='protocol',
            regex='regex',
            handler='handler',
            handler_type='handler_type'
        ),
    ]


    url_maps = super().register_url_maps(controller_kwargs)

    url_maps.append(UrlMap(...))
    return url_maps
```

app_base .py

```python
        from .controller import register_controllers

        controller_modules = [f'{self.package_namespace}.{self.package}.{module_name}' for module_name in
                              set(DEFAULT_CONTROLLER_MODULES + self.controller_modules)]
        return register_controllers(
            root_url=self.root_url,
            modules=controller_modules,
            index=self.index,
            catch_all=self.catch_all,
            extra_controller_kwargs=controller_kwargs,
        )
``` 

controllers.py

```python
def register_controllers(
        root_url: str,
        modules: Union[str, list, tuple],
        index: str = None,
        catch_all: str = '',
        extra_controller_kwargs: list = None,
) -> list:
    global app_controllers_list
    ...
    app_controllers_list = list()
    for module in all_modules:
        importlib.reload(module)  # load again to register controllers

    app_controllers_list.extend(extra_controller_kwargs)

    names = dict()
    for kwargs in app_controllers_list:
        name = kwargs['name']

    ...
```

Reply:
@swainn Sorry, I commented before refreshing and seeing your response. It seems like the easiest thing to do in your example would just be to have an option to disable the auto-register from the home controller.  The `register_controllers` already accepts an `index` argument. If it's `None` then it will skip over any of the home controller logic, so you could just do this:

```python
def register_url_maps(self):

    url_maps = super().register_url_maps(index=None)

    url_maps.extend(url_maps_from_extension)
    return url_maps
```

Would that work?

Reply:
I think so... I'm trying to remember how Tethys knows which URL to use as the home URL and what role the register_url_maps function has in that.